### ISSUE TITLE
.npz files visualization

### ISSUE THREAD

**Issue by dmetehan** [Arthur151/CenterHMR#15]
I would like to visualize the body landmarks projected onto the 2d image however I have some problems understanding how the values are stored. When I run the demo script I get a file called .npz and I can load it with numpy. Inside it I found f->results where two dictionaries are stored one for each person in the image. Inside each dictionary I can see 'cam' (3,), 'pose' (72,), 'betas' (10,), 'j3d_smpl24' (24, 3), 'j3d_op25' (25, 3), and 'verts' (6890, 3) variables (I put their corresponding shapes in parentheses). From your paper I thought 'pose' variable would be of length 132 (22 landmarks x 6D). Anyway, I assume you have saved 24 landmarks in 3D which makes an array of length 72. From the 'cam' variable (tx, ty = cam[1:]) I calculated the center for each people trivially since those numbers are normalized between -1 and 1 just as you mentioned in your paper. However when it comes to visualizing 'pose', 'j3d_smpl24' or 'verts' variables I ran into some problems. Could you explain how each of these variables store their data? Apparently they are not normalized between -1 and 1. Also I'm having some problems understanding the first number in 'cam' variable which should correspond to scale according to the paper. In the paper it is said that this variable "reflects the size and depth of the human body to some extent.". How can this scale be used to visualize pose points? In my example I get 5.51 for one person and -5.156 for the other. Would you also explain what does a negative scale represent?

**Comment by Arthur151** [comment]
Good question! 
1.About the 3D-to-2D projection. It is presented at [here](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/core/base.py#L84). The detailed function is at [here](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/utils/projection.py#L16). To get the projected 2D coordinates, just use 'pj2d' at [here](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/core/base.py#L87). They are normalized (-1~1) coordinates on the input image (not  the original image). 
2. About 'cam'. We adopt a 3-dim camera parameter. You can treat the absolute value of the first dim. as the scale. These camera parameters are used to project the estimated 3D joints or body vertex back to 2D image via [weak-perspective projection](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/utils/projection.py#L16). 
3. About 'pose': they are 72-dim SMPL pose parameters, which are 3-dim 3D rotation of 24 SMPL joints 'j3d_smpl24' in axis representation.
4. About 'j3d_smpl24' or 'verts'. They are in standard SMPL space, not image space. In standard SMPL space, the body center (near pelvis) is located at the origin.

Don't hestiate to let me your question.
Best.

**Comment by dmetehan** [comment]
Thank you very much for the detailed info. Is there a way to map j3d_smpl24 onto the image space?

**Comment by Arthur151** [comment]
Of course. It is very simple. Add a line after [this line](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/core/base.py#L84),
`pj2d_smpl24 = proj.batch_orth_proj(j3d_smpl24, params_dict['cam'], mode='2d')[:,:,:2]`
'pj2d_smpl24' is what you want. Please add it to the 'outputs' dict for further use.

**Comment by lisa676** [comment]
@Arthur151  Hi dear Yu Sun. I'm also trying to get SMPL 3D keypoints in image space. I followed your instructions and I got 2D data and 45 entries. I'm writing it as example below. 
`[[0.0744452346641, -0.012984120034], [-0.23053123445656, 0.00345313533],.... upto 45 entries]` 
I want to ask that how to get 3D keypoints in image space? 

One more question as you mentioned above reply that in SMPL space body center is pelvis. So what about body center in image space?

**Comment by Arthur151** [comment]
1. Please set keep_dim=True to keep the Z dimention. Please refer to [projection function](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/utils/projection.py#L16). 
2. What do you mean by "body center in image space"? Are you asking how to determine the body center in center map? We have detailed description about this in our paper. Briefly speaking, 2D body center could be regarded as the average of torso keypoints (left/right shoulders, left/right hips and so on).
3. I haven't looked into the "SPIN keypoints" you mentioned. But I think it can be derived from SMPL body mesh via their regression functions.

**Comment by lisa676** [comment]
> Good question!
> 1.About the 3D-to-2D projection. It is presented at [here](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/core/base.py#L84). The detailed function is at [here](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/utils/projection.py#L16). To get the projected 2D coordinates, just use 'pj2d' at [here](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/core/base.py#L87). They are normalized (-1~1) coordinates on the input image (not the original image).
> 2. About 'cam'. We adopt a 3-dim camera parameter. We don't directly use the estimated scale value. Instead, we take the [$(1.1)^scale$](https://github.com/Arthur151/CenterHMR/blob/00c43337855c41fffae7ea552dda9be05cacd282/src/utils/projection.py#L18) to make sure that the scale value is always positive. These camera parameters are used to project the estimated 3D joints or body vertex back to 2D image via [weak-perspective projection](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/utils/projection.py#L16).
> 3. About 'pose': they are 72-dim SMPL pose parameters, which are 3-dim 3D rotation of 24 SMPL joints 'j3d_smpl24' in axis representation.
> 4. About 'j3d_smpl24' or 'verts'. They are in standard SMPL space, not image space. In standard SMPL space, the body center (near pelvis) is located at the origin.
> 
> Don't hestiate to let me your question.
> Best.

Thanks for your reply. In this reply in point 4 you mentioned that _In standard SMPL space, the body center (near pelvis) is located at the origin._  I was asking about this body center. 

Furthermore I changed `keep_dim=True` but still I'm getting same results as I mentioned above.

**Comment by Arthur151** [comment]
1. After projection, there may not be any center located at the origin. Because there are X-/Y-translation.
2. Because we have forced the 'pj2d' to be 2-dims in [this line](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/core/base.py#L87). As we can see, we set 'pj2d':pj3d[:,:,:2]. If you want to get the 3-dims, please change it to 'pj2d':pj3d.

**Comment by lisa676** [comment]
Thanks for clarification. Now I can get x, y and z data by removing `[:,:,:2]` but there is one confusion that there is 45 entries ` array size is (1, 45, 3)` for one frame but I think it should be `1, 24, 3`.

**Comment by Arthur151** [comment]
To get the back-projected 3D joints in SMPL format, I recommand to replace [this line](https://github.com/Arthur151/CenterHMR/blob/81af7e881db80f9d4854d7cfdcdad472b90556fb/src/core/base.py#L84), which is 
`pj3d = proj.batch_orth_proj(j3d_op25, params_dict['cam'], mode='2d')`
with 
`pj3d = proj.batch_orth_proj( j3d_smpl24, params_dict['cam'], keep_dim=True)`

**Comment by lisa676** [comment]
@Arthur151  Thanks for your help. I'm following your suggestions but still I'm getting `(1, 45, 3)`. I don't know why it is happening? Even I removed `j3d_op25` as you suggested above. Would you like to test this problem on your side?

**Comment by Arthur151** [comment]
These 3D body joints are derived from the estimated SMPL body mesh. Each body mesh contains 6790 vertices, from where we can regress the 24 body joints you need. The regression process is quite simple. Because these vertices have stable semantic location, each joint could be easily regressed from them. For instance, if we want to calculate the elbow keypoints, we make average on the vertices near the elbow. In this way, theoritically, we can derive any keypoint you want from the body mesh. There might be some redundant keypoints we regressed from the SMPL body mesh, which makes it 45 instead of 24.
